Publish ES5 sources to NPM, get react-native-web compat for free

emirotin opened the pull request:
I notice it's common for RN libs to be published in ES6+ and often with flow types. While this is fine with RN itself it's uncommon for the web-targeting modules which are supposed to be published in the common denominator format.

Specifically if one is using create-react-app which is very popular of course node_modules are not processed by babel and are only packaged by webpack. So certain features like flow types, or spread destructuring in case of this module make the build break.

Now why would one need RN lib in the web / with CRA? Because of React Native for web project.
I'm currently working on the Electron app with RN-web and would like to later reuse the same components for the actual RN mobile app.

So I suggest following the common pattern of publishing the compiles ES5 sources to NPM. I've also slightly updated the dependencies and simplified the Prettier+ESLint config following the recommendations from the Prettier docs.

The PR currently contains a bunch of garbage commits (fix xyz) that I will squash before merging if you approve the general suggestion.

I've also tested my fork with my current project and it works:
![Alt text](https://monosnap.com/file/mmvTKXRqvytJKDkm32c1dX7aYXE8ro.png)

emirotin commented:
I will also put some comments/annotations to the sources now.

emirotin commented on .babelrc:
```diff
@@ -1,3 +1,14 @@
-  "presets": ["react-native"]
+  "presets": [
+    [
+      "env",
```
This is the modern replacement fro es2015 preset. It autodetects the features to be transpiled / polyfilled based on the target platforms.

emirotin commented on .eslintignore:
```diff
@@ -1,3 +1,4 @@
-example
\ No newline at end of file
+example
+lib
```
Built sources should not be linted

emirotin commented on .eslintrc.js:
```diff
@@ -1,58 +1,29 @@
-
 module.exports = {
-  "extends": ["eslint:recommended", "plugin:react/recommended"],
+  extends: ['eslint:recommended', 'plugin:react/recommended', 'prettier'],
```
All the formatting was done by Prettier. It can be restored and file added to ignores if you wish.

emirotin commented on .eslintrc.js:
```diff
@@ -1,58 +1,29 @@
-    "no-console": 0,
-    "no-shadow": 2,
-    "no-var": 2,
-    // style
```
No need for style rules — we're using Prettier so can rely on its own rules set. The `'prettier/prettier': 'error',` ensures linting will fail if there's contradiction with prettier's formatting.

emirotin commented on __tests__/HTMLView-test.js:
```diff
@@ -4,7 +4,7 @@ import renderer from 'react-test-renderer';
 import {StyleSheet, Text} from 'react-native';
 
-import HTMLView from '../HTMLView';
+import HTMLView from '../lib';
```
Test the built sources instead of originals to also capture possible transpilation issues

emirotin commented on example/Example.js:
```diff
@@ -1,14 +1,12 @@
 import React from 'react';
 import {StyleSheet, View, Text, ScrollView} from 'react-native';
-import HTMLView from '../';
+import HTMLView from '../src/';
```
Example can keep using sources

emirotin commented on format.sh:
```diff
@@ -1,12 +0,0 @@
-  echo "Error: expected args"
-  exit 1
-fi
-node_modules/.bin/prettier \
```
Moved the common prettier config to package.json so no need for this file anymore

emirotin commented on package.json:
```diff
@@ -2,23 +2,31 @@
   "version": "0.12.1",
   "description": "A component which renders HTML content as native views",
-  "main": "index.js",
+  "jsnext:main": "src/index.js",
```
This allows for original ES6 sources to be used by compatible packagers

emirotin commented on package.json:
```diff
@@ -2,23 +2,31 @@
   "description": "A component which renders HTML content as native views",
-  "main": "index.js",
+  "jsnext:main": "src/index.js",
+  "main": "lib/index.js",
```
This is for "regular" consumers

emirotin commented on package.json:
```diff
@@ -2,23 +2,31 @@
+      "yarn run format-lint && yarn run lint && yarn run babel && yarn run jest",
+    "format-lint":
+      "prettier --list-different \"./**/*.js\"; if [ $? != 0 ]; then echo \"CODE FORMATTING: please run 'yarn run format' and commit the changes\"; exit 1; fi",
+    "format": "prettier --write \"./**/*.js\" && eslint . --fix",
```
We can actually drop the `eslint` part now

emirotin commented on package.json:
```diff
@@ -2,23 +2,31 @@
-    "format": "./format.sh --write && eslint . --fix",
-    "jest": "jest"
+    "test":
+      "yarn run format-lint && yarn run lint && yarn run babel && yarn run jest",
```
Build sources before testing

emirotin commented on package.json:
```diff
@@ -2,23 +2,31 @@
+    "format": "prettier --write \"./**/*.js\" && eslint . --fix",
+    "jest": "jest",
+    "babel": "babel src --out-dir lib --source-maps",
+    "prepublishOnly": "yarn run test"
```
Run tests (and build sources) before publishing

emirotin commented on src/AutoSizedImage.js:
```diff
@@ -1,8 +1,5 @@
 import React, {PureComponent} from 'react';
```
All sources moved to a separate folder to separate them from repo spam and simplify the babel command

richchurcher commented:
Hi @emirotin ! Thanks for your thorough reworking of the build, lint and test process, and for your patience. There are a few moving parts in this PR and I want to sit with it for just a bit longer before I merge :smile: I will take another look end of day today or tomorrow.

People wondering about transpilation of RN modules may be interested in https://github.com/facebook/react-native/issues/10966 .

emirotin commented:
Hey, I know it's been a while but I've now returned to the project I was using your lib for and with all the deps updated (react, react-native-web, react-scripts, etc.) it's still the same issue. So I've updated this PR, rebased it against your most recent master and squashed the technical commits. The thing works for me as expected so I suggest you reconsider this PR.
One thing that can be changed if you wish is excluding the `lib/` from git as it's the derived artifact that can always be built from sources before testing and before publishing to NPM. I need it right now though because I'm using this module directly from my git fork and I need it usable without the build step.

emirotin commented on .prettierignore:
```diff
@@ -0,0 +1 @@
+lib
```
Without this, the source maps would be broken.
